### ISSUE TITLE
Unexpected behavior in list reconciliation

### ISSUE THREAD

**Issue by bluenote10** [ryansolid/solid#51]
I'm not sure what is a minimal reproducing example of this issue, but I can reproduce it as follows ([codesandbox](https://codesandbox.io/s/debugforwithindex-9u44u)):

```js
import { createState } from "solid-js";
import { render, For } from "solid-js/dom";

const App = () => {
  const [state, setState] = createState({
    data: ["a", "b", "c"],
    selectedIndex: -1
  });

  setTimeout(() => {
    setState({
      selectedIndex: 1
    });
  }, 1000);

  setTimeout(() => {
    setState({
      data: ["b"],
      selectedIndex: 0
    });
  }, 2000);

  /* prettier-ignore */
  return (
    <div>
      <div>Selected Index: {(state.selectedIndex)}</div>
      <For each={(state.data)}>{ (x, i) => 
        <div class={(i === state.selectedIndex ? "selected" : "")}>
          {x}
        </div>
      }</For>
    </div>
  );
};

render(App, document.getElementById("app"));
```

Behavior:
- at t = 1 sec the selected index is set to 1, selecting element `b` (visible border).
- at t = 2 sec the data updates to just `["b"]` adjusting the selected index to 0. Thus, element `b` should still have a border, but it doesn't.

This issue might by related to #50, because it is also caused by having identical nodes: If the data update at t = 2 sec sets the data to `["x"]`, the item gets selected properly.

I have experienced this issue in a "search as you type" component with item selection (that's why narrowing down the item list is the typical behavior). In my app there can be all sorts of funky states, including off-by-one selections, or multiple selected items. I'm not sure if these are multiple issues, or just manifestations of the same problem. Or if I'm doing something wrong ;).

----

Side note: The TypeScript definitions of `<For>` currently do not allow functions of the type `(item: T, i: number)`, but the underlying `map` function does. This allowed me to define a `<ForIndexed>` on user site. It's really nice that the new control flow elements are fully accessible compared to `<$>` and can easily extended on user site.

**Comment by ryansolid** [comment]
I see what your issue is. I remember now why I might have not put the index on the `<For>` component. Truth is we probably could but I think it would lead to this confusion a lot. The problem is that `i` is not a signal. It is value wrapped in a closure. So for `b`, `i` will always be 1 in your example. Just because indexes have changed I do not re-run every template. That would be extremely inefficient (recreating nodes and such).

Ironically it was this scenario also that was the final push for me to change to this control flow. Generally I do everything with lists with objects and ids. That's basically the difference with React. We reconcile in the data while they reconcile in a virtual DOM representation. 

If you really want to use indexes, what you want is to make your indexes signals. I actually already have an example for just that. https://codesandbox.io/s/list-indices-w3vlv. This also has that flip animation library but you can see how I use map twice to add signal indexes to row.

**Comment by bluenote10** [comment]
Cool, the `<ForIndex>` works nicely, thanks! Would it make sense to include it in Solid? Here's a first conversion to TypeScript:

```ts
interface MergedGetterSetter<T> {
  (): T;
  set(x: T): void
}

export function ForIndex<T, U>(props: {
    each: T[],
    children: (item: T, n: () => number) => U,
  }) {
  const fn = props.children;
  const list = () => props.each;
  // We createMemo to prevent work being done each row in the final map we return
  const indices = createMemo(
    map((x: T, i: number) => {
      const [s, set] = createSignal(i, (a, b) => a === b) as [MergedGetterSetter<number>, (i: number) => void];
      s.set = set;
      return s;
    })(list)
  );
  createEffect(() =>
    indices().forEach((index, i) => index.set(i))
  );
  return map((v: T, i: number) => fn(v, sample(indices)[i]))(list);
}
```
Note: I'm a bit confused that the map for the indices uses `map(i => ...)`. What is `i` here? The list passed into the map is of type `T[]` so the first argument must be a `T`. The intention of the created signals is to be of type `number` though. I have modified the code to use `map((x: T, i: number) => ...` and just ignore the `x` which makes everything consistent. Interestingly, both versions seem to work.

-----

As a beginner, it is a bit non-intuitive why the `x` and the `i` provided in my initial attempt have different behavior. As I understand it, the difference is that `x` originates from proxied data (and thus is observable / a signal), whereas the `i` is just a plain number.

For TypeScript users there is probably no issue here, because they cannot accidentally pass in a `(x: T, i: number)` function to `<For>` and the compiler also makes them aware of using `i()` in the `<ForIndex>` version. To mitigate the problem for JS users, would it make sense to use proxies for the index so that writing just `i` is possible, or would that be less efficient that signals?

**Comment by ryansolid** [comment]
Proxies have to be objects. And I intercept property access. `x` is just as static as `i`. It's the property access on `x` or nested signals that allow dynamic updates. I currently in favour of function getters on signals over Vue RFC style `i.value` as it allows passing around easier rather than always wrapping like we do with state, not to mention slightly worse performance. It is a fair thing to be up for discussion though.

Yes the way you corrected it is right. It was getting set again by the effect anyway so I didn't notice. What I wrote was a bug.

I have no idea why `indices` ever be undefined. Do you mean the specific index or the whole `indices` array? They are both triggering off the same array using the same map function.

**Comment by bluenote10** [comment]
It took me a while to narrow down the remaining issue I'm having with `<ForIndex>`. It is caused by having a `<ForIndex>` within a `<Switch>`/`<Match>` ([codesandbox](https://codesandbox.io/s/debugforindex-5f073)):

```js
const App = () => {
  const [state, setState] = createState({
    data: ["a", "b", "c"]
  });

  /* prettier-ignore */
  return (
    <div>
      <div>
        <Switch>
          <Match when={true}>
            <ForIndex each={(state.data)}>{ (x, i) => 
              <div>
                {x}
              </div>
            }</ForIndex>
          </Match>
        </Switch>
      </div>
    </div>
  );
};
```

The example works fine when removing the `<Switch>`/`<Match>`. For some reason the computation coming out of `createMemo` gets messed up.

**Comment by ryansolid** [comment]
Yeah Switch/Match is unlike the others. To keep it light I didn't use Context API maybe when I should have. I suspect that will be our problem here. I will see if I can confirm.

**Comment by ryansolid** [comment]
Ok I figured it out. I broke synchronicity in createMemo a while back trying to simplify adding comparators. It's causing an actual clock cycle to happen where the value initially is undefined. I need to fix this as it causes inefficiency in some cases. Short term if we just initialize the memo with an empty [] we are fine but that isn't the long term solution. I updated my example, but this is not the long term solution here. I can see other scenarios where this could be problematic.

**Comment by bluenote10** [comment]
Thanks for the workaround! In my app I'm running into more issues though, maybe they are related. Specifically if I call `i()` in an event handler it returns undefined instead of a number.

I tried reproducing this problem [here](https://codesandbox.io/s/debugforindex-5f073) with this code:

```js
const App = () => {
  const [state, setState] = createState({
    data: ["a", "b", "c"]
  });

  /* prettier-ignore */
  return (
    <div>
      <Switch>
        <Match when={true}>
          <ForIndex each={(state.data)}>{ (x, i) => 
            <a onclick={ event => console.log("clicked:", i()) }>
              <div>{x}</div>
            </a>
          }</ForIndex>
        </Match>
      </Switch>
    </div>
  );
};
```

But in this case the problem manifests itself differently: The event handler never get bind at all. Again, if I remove the `<Switch>`/`<Match>` the problem disappears. Hopefully that all has the same cause.

**Comment by ryansolid** [comment]
Looks to be the same issue. Fixed v0.9.4. Let me know if you see anymore strangeness.

**Comment by bluenote10** [comment]
Thanks for the quick fix!

I didn't had much time to check it, but in my app some of the update inconsistencies have disappeared now, some remain. I need to take a closer look this weekend, I hope I can make sense of it.

I saw that you have extended `<For>` taking an index. Does that mean `<ForIndex>` now isn't needed?

**Comment by ryansolid** [comment]
No.. it's a number. ForIndex is still needed for shifting indexes.. That was my slip up. I was trying to tighten up the types a bit better looking at your pull request and I must have accidently added that in.

Hmm.. it might still be synchronicity issues. I'm really hoping not. When I did the refactor for the RFC #42 I changed my solution for If and Switch to be non-synchronous. I thought I was handling it well enough for those specific cases as long as I made the library synchronous in general. It's just really hard to do a simple map when the conditions can evaluate several times to the same value and not expect an update.

Like Show if  x > 5, will retrigger (releasing child deps) for x = 6, x = 7, x = 8. It's really wasted when all you want is to run the code when the condition. My options are:
1. Manage it under a root like I used to being conscious of garbage collection (previous solution)
2. Hoist the condition outside of the computation but then it's always rendered (doesn't really work if the condition is there because say the data is missing).
3. Break synchronicity by storing intermediate results in a signal that uses a comparator (what I'm doing now).
4. Use some mechanism to allow comparators on `createMemo` to decide to whether to notify. Which is an interesting topic which seems to be a research one for S.js library. Adam has been working on Sub Clocks and this post has an alternative approach: https://github.com/adamhaile/S/issues/26.

**Comment by ryansolid** [comment]
I have confirmed there still are. I need to remove all the broken synchronicity as a first step. And it has come to my attention directly nested multi children control flows can have some awkward inefficiencies. Fixing these is my highest priority for the library right now.

**Comment by bluenote10** [comment]
:+1:

Good news from my side: I could resolve all the issues I had now. Turned out they were simply caused by a misunderstanding when `forwardRef` triggers ([see reddit question](https://www.reddit.com/r/solidjs/comments/ckui1o/is_it_safe_to_createeffect_in_forwardref/)). Good to hear though that you have reached a really deep understanding of all the synchronicity edge cases now.

**Comment by ryansolid** [comment]
Yeah with the fix to all synchronicity in v0.9.5. I think this issue has run its course. Thanks for all the examples. It really makes it a lot easier.
